### DISCUSSION TITLE
Scenario Test Usecase (no events, number of events)

### DISCUSSION THREAD

**Discussion by wolframhaussig** [spring-projects/spring-modulith#589] [Q&A] [answered]
Dear all,

I am trying to create unittests for my Spring Modulith application. Creating tests for published events are easy:
```
scenario.stimulate(() -> collector.poll())
                .andWaitForEventOfType(NewJobEvent.class)
                .toArrive();
```

What  I fail to create:

1. A test that checks that an event was **NOT** published. The usecase I try to replicate is that the publisher can be configured to temporarily ignore specific event types (e.g. because the target system is not available). So I am looking for something like this:
```
scenario.stimulate(() -> collector.poll())
            .andWaitAtMost(Duration.ofSeconds(5))
            .andWaitForEventOfType(NewJobEvent.class)
            .not().toArrive();
```

2. A test that checks for a specific number of events being published. The usecase here is that it should only publish events until a given limit of currently executing jobs is reached. This limit is used to prevent one instance of the service to block all jobs while another instance is idle.
```
scenario.stimulate(() -> collector.poll())
            .andWaitAtMost(Duration.ofSeconds(5))
            .andWaitForEventOfType(NewJobEvent.class)
           .count(3)
            .toArrive();
```

Any ideas how to replicate something like this?

Best regards
Wolfram

**Comment by odrotbohm** [comment, marked as answer]
Can you please go ahead and create an issue for the first case? I am not sure yet whether we should go with that intermediate `….not()` or additional `….notToArrive()`, as some overloads of `….toArrive(…)` take the event we're expecting in some form or another, which doesn't work if there's no event to expect. The latter approach, however, might add quite a few new methods, which is not without problems either.

Can you also please do the same for the second item? I guess I'd rather like to integrate that into the `….toArriveAndAssert(…)` flavor, expanding on the `PublishedEventAssert` we already have in place.

**Reply by wolframhaussig** [reply]
I created #1357 and #1358 for this
